core/src/test/java/org/apache/iceberg/rest/TestHTTPRequest.java

```diff
@@ -0,0 +1,138 @@
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+class TestHTTPRequest {
+
+  @ParameterizedTest
+  @MethodSource("validRequestUris")
+  public void requestUriSuccess(HTTPRequest request, URI expected) {
+    assertThat(request.requestUri()).isEqualTo(expected);
+  }
+
+  public static Stream<Arguments> validRequestUris() {
+    return Stream.of(
+        Arguments.of(
+            ImmutableHTTPRequest.builder()
+                .baseUri(URI.create("http://localhost:8080/foo"))
+                .method(HTTPRequest.HTTPMethod.GET)
+                .path("v1/namespaces/ns/tables/") // trailing slash should be removed
+                .putQueryParameter("pageToken", "1234")
+                .putQueryParameter("pageSize", "10")
+                .build(),
+            URI.create(
+                "http://localhost:8080/foo/v1/namespaces/ns/tables?pageToken=1234&pageSize=10")),
+        Arguments.of(
+            ImmutableHTTPRequest.builder()
+                .baseUri(URI.create("http://localhost:8080/foo"))
+                .method(HTTPRequest.HTTPMethod.GET)
+                .path("https://authserver.com/token") // absolute path HTTPS
+                .build(),
+            URI.create("https://authserver.com/token")),
+        Arguments.of(
+            ImmutableHTTPRequest.builder()
+                .baseUri(URI.create("http://localhost:8080/foo"))
+                .method(HTTPRequest.HTTPMethod.GET)
+                .path("http://authserver.com/token") // absolute path HTTP
+                .build(),
+            URI.create("http://authserver.com/token")));
+  }
+
+  @Test
+  public void malformedPath() {
+    assertThatThrownBy(
+            () ->
+                ImmutableHTTPRequest.builder()
+                    .baseUri(URI.create("http://localhost"))
+                    .method(HTTPRequest.HTTPMethod.GET)
+                    .path("/v1/namespaces") // wrong leading slash
+                    .build())
+        .isInstanceOf(RESTException.class)
+        .hasMessage(
+            "Received a malformed path for a REST request: /v1/namespaces. Paths should not start with /");
+  }
+
+  @Test
+  public void invalidPath() {
+    HTTPRequest request =
+        ImmutableHTTPRequest.builder()
+            .baseUri(URI.create("http://localhost"))
+            .method(HTTPRequest.HTTPMethod.GET)
+            .path(" not a valid path") // wrong path
+            .build();
+    assertThatThrownBy(request::requestUri)
+        .isInstanceOf(RESTException.class)
+        .hasMessage(
+            "Failed to create request URI from base http://localhost/ not a valid path, params {}");
+  }
+
+  @ParameterizedTest
+  @MethodSource("validRequestBodies")
+  public void encodedBody(HTTPRequest request, String expected) {
+    assertThat(request.encodedBody()).isEqualTo(expected);
```

Comment:
personally I think the test code would be much more readable without having a parameterized test and rather have just two separate test methods that verify the body is properly encoded as form data/JSON, but I'm ok with what's currently here and I'll leave that up to you.

Also we probably might want to have some tests where the body can't be encoded due to:
* body is null
* body is a Map with null values
* body is an invalid JSON